MSCXXXX: Fixes for Client Well-known URI (#2499)
matrix-org/matrix-doc: proposals/2499-client-well-known-fixes.md

# MSC2499: Fixes for Well-Known URIs

Several minor issues and inconsistencies have come up with the well-known URIs used for
discovering clients and servers. This proposal intends to fix these issues.

## Proposal

The following changes should be made to the spec:

1. The spec does not mention that redirects should be followed for `/.well-known/matrix/client`
and does not clearly specify what type of redirects should be followed for `/.well-known/matrix/server`.
To fix this, the spec should be changed to state that when a well-known URI is requested,

> the client or server should follow HTTP 301, 302, 303, 307, and 308 redirects up to 30 times.

Reviewer: Should a redirect to a `http://` URL be allowed? (for security reasons)

Reviewer 2: Clients don't really have reasonable control over this, so would recommend leaving that as an implementation detail.

1. Currently the spec does not mention which Content-Type should be used for the response to
`/.well-known/matrix/client`. The Client-Server spec should be changed to state the Content-Type
SHOULD be `application/json` however it should be assumed to be JSON regardless of Content-Type.
This is consistent with the Server-Server API.

> 1. Step 3f in the Client-Server well-known flow should be changed to use the modern

Reviewer 3: ```suggestion
1. Step 3.6 in the Client-Server well-known flow should be changed to use the modern
```

The latest rendering of the spec uses numbers for the sub-steps instead of letters

`/_matrix/identity/versions` API to validate the identity server rather than the deprecated
`/_matrix/identity/api/v1`.

1. Mark the `/_matrix/identity/v2` status check API as deprecated. `/_matrix/identity/versions`
is the suggested replacement.

> 1. The maximum size of size of the well-known file is 51200 bytes. A client or server

Reviewer 4: ```suggestion
1. The maximum size of the well-known file is 51200 bytes. A client or server
```

> requesting a well-known file MUST abort and FAIL_PROMPT if the response exceeds 51200 bytes.

Reviewer 3: ```suggestion
requesting a well-known file MUST abort and FAIL_PROMPT if the response exceeds 51200 bytes.
This check becomes part of step 3.2.
```

> # MSC2499: Fixes for Client Well-known URI

Reviewer 5: (If this should be discussed elsewhere please advise)

Should clients also pre check the file size? This is also a thing I especially saw on SS API happening where it tries endpoints that are up to 5GB or more. This could probably be circumvented by checking the Content-Length before fetching the content and rejecting it after a specific size.

Not sure if it belongs in this MSC though. Feel free to move this idea elsewhere.

Author: I'd like to include any fixes that need to happen to well-known URIs in this MSC so they can all be fixed at once so here is fine.

That makes sense to me. Synapse currently limits the server well-known file to 50KB https://github.com/matrix-org/synapse/pull/8950. Does that seem like a reasonable limit to you? Should this be a MUST NOT exceed X size type of thing or should it be just a suggestion?

Reviewer 5: I would prefer actually a "MUST NOT" because a suggestion could cause a) that it just doesnt get followed b) there exist various different interpretations c) Clients or Servers would behave different which would add more confusion to users.

But I think that wording might be better decided by SCT

Several minor issues and inconsistencies have cropped up since the `/.well-known/matrix/client`
method was added to allow clients to discover servers. This proposal intends to fix these issues.

## Proposal

The following changes should be made to the client server spec:

1. The spec does not mention that redirects should be followed. The spec should be changed to

> state that when a client is looking up the well-known URI, it should follow HTTP 3xx redirects

Reviewer 6: we recently discovered that synapse wasn't following 308 redirects. We might want to enumerate the valid redirect codes.

Author: Yes I think that is a good idea to make sure it is clear specifically what is acceptable. That will also prevent people from reading "30x redirects" to mean follow up to thirty redirects.

Do you have a list of codes that should be valid? https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#3xx_redirection

300, 301, 302, 303, 304, 307, 308 I would guess? Probably not 305 or 306.

Reviewer 6: 300 and 304 aren't redirects afaict. so: 301, 302, 303, 307, 308.

up to 30 times. This is consistent with the redirects followed when a server resolves a
server name using `/.well-known/matrix/server`.

1. Currently the spec does not mention which Content-Type should be used for the response to
`/.well-known/matrix/client`. The Content-Type SHOULD be `application/json` however it
should be assumed to be JSON regardless of type. This is consistent with the Server-Server API.

1. When a client is following the flow described in the spec for looking up `/.well-known/matrix/client`,
the spec currently states, in part:

    > 3․ Make a GET request to https://hostname/.well-known/matrix/client.<br>
	>>  a․ If the returned status code is 404, then `IGNORE`.<br>
    >>  b․ If the returned status code is not 200, or the response body is empty, then `FAIL_PROMPT`.

    In practice most web servers do not add CORS headers on 404 errors by default. Therefore
    web based clients cannot always determine if the status code is 404. Step 3b should be 
    changed from `FAIL_PROMPT` to `IGNORE` so that a non 200 response is treated in the same
    way as 404. This change is intended to fix issues like https://github.com/vector-im/riot-web/issues/7875.
  

>     This change does have potential security concerns, see https://github.com/vector-im/riot-web/issues/11136.

Author: You know what, actually reading the spec again maybe FAIL_PROMPT is the correct thing to do but Riot just isn't doing that. Riot is doing something closer to a FAIL_ERROR, rather than a FAIL_PROMPT. I'm thinking IGNORE should just be removed from the spec completely and 3a and 3b should be a FAIL_PROMPT.

>     This change does have potential security concerns, see https://github.com/vector-im/riot-web/issues/11136.

Reviewer 7: I wouldn't be so harsh on IGNORE - it allows the user to go without an unneeded roadbump in 3a in case the client managed to get the clear message from the server that there's no .well-known support here. The IGNORE description is too affording though.

>     This change does have potential security concerns, see https://github.com/vector-im/riot-web/issues/11136.

Author: Yeah it would be nice if there was no extra step but if using defaults values is removed from IGNORE then how will the client find out the server URL? The only way is to prompt the user unless I am missing something. So it would be the same as FAIL_PROMPT.

>     This change does have potential security concerns, see https://github.com/vector-im/riot-web/issues/11136.

Reviewer 7: Probably the spec should be clear about the scope of cases for auto-discovery. Specifically, if the user entered a fully qualified MXID, only then auto-discovery kicks in. Using the default value in a situation when the serverpart is already provided but the server happens to have .well-known not quite ideally configured (i.e. step 3b) is definitely the wrong way out. Asking the user to confirm that the server is correct because auto-discovery fell short is the right way but both IGNORE and FAIL_PROMPT speak about that anyway.

>     This change does have potential security concerns, see https://github.com/vector-im/riot-web/issues/11136.

Reviewer 2: `FAIL_PROMPT` is probably still correct, I think. Element (then Riot) likely has a bug in this area, though its UX doesn't really lend itself to "prompting" the user. It should probably offer the ability to continue anyways, though.

>     This change does have potential security concerns, see https://github.com/vector-im/riot-web/issues/11136.

Author: Yeah I'll take this out.

`/.well-known/matrix/client`. The Client-Server spec should be changed to state the Content-Type
SHOULD be `application/json` however it should be assumed to be JSON regardless of Content-Type.
This is consistent with the Server-Server API.

1. The Client-Server spec for looking up `/.well-known/matrix/client` currently states, in part:

    > 3․ Make a GET request to https://hostname/.well-known/matrix/client.<br>
	>>  a․ If the returned status code is 404, then `IGNORE`.<br>
    >>  b․ If the returned status code is not 200, or the response body is empty, then `FAIL_PROMPT`.

    In practice most web servers do not add CORS headers on 404 errors by default. Therefore
    web based clients cannot always determine if the status code is 404. Step 3b should be 
    changed from `FAIL_PROMPT` to `IGNORE` so that a non 200 response is treated in the same
    way as 404. This change is intended to fix issues like https://github.com/vector-im/riot-web/issues/7875.
  
    This change does have potential security concerns, see https://github.com/vector-im/riot-web/issues/11136.

1. Step 3f in the Client-Server well-known flow should be changed to use the modern

> `/_matrix/identity/v2` API to validate the identity server rather than the deprecated

Reviewer 2: ```suggestion
`/_matrix/identity/versions` API to validate the identity server rather than the deprecated
```

Author: Didn't you just spec this in https://github.com/matrix-org/matrix-doc/pull/3170/ as `/_matrix/identity/v2`?

Reviewer 2: I did, but that's because it's a direct replacement and in line with the proposals. For better or worse, we can't change intention at the spec PR stage but we can ask that other MSCs (or even new ones) to do the right thing.

Overall, the `/versions` endpoint (which is new since this MSC was opened) is better suited for long-term success of the spec. Bonus points for deprecating the v2 ping endpoint, but not necessary to fix the issue.

`/_matrix/identity/api/v1`.

1. The maximum size of size of the well-known file is 51200 bytes. A client or server

> requesting a well-known file MUST abort and FAIL_PROMPT if the response exceeds 51200 bytes.

Reviewer 2: This isn't a super realistic thing that can be applied by clients: if the server admin has configured the server so badly that a 10gb JSON file is being served then so be it. It's just not in the server admin's interest to attack bandwidth in this way given the costs they would be incurring from their ISP/service provider.

Reviewer 7: At least for native clients (and servers) it's fairly easy to abort on 51201st byte but `FAIL_PROMPT` looks a bit dubious here: should the client re-request .well-known or rather just continue (rather than abort) receiving the payload after confirming with the user? And yes, generally there's no incentive to attack in that way of course; but a recommendation to protect receiving sides from such misconfigurations wouldn't hurt, would it? Maybe not as MUST but as SHOULD.

Reviewer 2: I'm not convinced that this endpoint specifically needs this recommendation though: surely we should be applying this to all endpoints like profiles, sync, etc... 

It feels like another MSC's problem.

Author: That's fine, I can leave it to another MSC. The main reason for including it here is because Synapse has already implemented this limit for the server well-known so a well-known bigger than that already isn't going to work with most servers. If it was specced then at least people could expect there is some reasonable limit on the size.

Reviewer 2: I would be a bit surprised if people expected 1mb json files to work for this use case, honestly. I think it's reasonable for clients/servers to impose their own limits, and iirc Synapse only really limited the JSON file size on server-server because federation has tiny timeouts (typically), so read delays are not ideal.
